File: command/command/commands/AbstractCommand.cs
using System;

namespace command.commands
{
	public abstract class AbstractCommand : IDisposable
	{
		private bool _isDisposed = false;

		public void Execute()
		{
			if (IsExecuted)
			{
				throw new InvalidOperationException("attempt to execute already executed command");
			}
			ExecuteImpl();
			IsExecuted = true;
		}

		public void Unexecute()
		{
			if (!IsExecuted)
			{
				throw new InvalidOperationException("attempt to unexecute already unexecuted command");
			}
			UnexecuteImpl();
			IsExecuted = false;
		}

		public void Dispose()
		{
			if (_isDisposed)
			{
				return;
			}

			DisposeImpl();
			_isDisposed = true;
		}

		protected abstract void ExecuteImpl();
		protected abstract void UnexecuteImpl();
		protected virtual void DisposeImpl() {}

		public virtual bool IsExecuted { get; protected set; } = false;
	}
}

File: command_java/Document/src/hsitory/AbstractCommand.java
package hsitory;

public abstract class AbstractCommand implements ICommand {

    @Override
    public final void Execute() {
        if (BeforeExecute())
        {
            ExecuteImpl();
            AfterExecute();
        }
    }

    @Override
    public final void UnExecute() {
        if (BeforeUnExecute())
        {
            UnExecuteImpl();
            AfterUnExecute();
        }
    }

    @Override
    public final void Destroy() throws Exception {

        if (BeforeDestroy())
        {
            DestroyImpl();
            AfterDestroy();
        }
    }

    @Override
    public final boolean IsExecuted() {
        return _ieExecuted;
    }

    protected boolean BeforeExecute(){
        if (_ieExecuted && _isAlive)
        {
            throw new IllegalStateException("attempt to execute already executed command");
        }
        return _isAlive;
    }
    protected void AfterExecute(){ _ieExecuted = true; }
    protected abstract void ExecuteImpl();

    protected boolean BeforeUnExecute(){
        if (!_ieExecuted && _isAlive)
        {
            throw new IllegalStateException("attempt to unexecute already unexecuted command");
        }
        return _isAlive;
    }
    protected void AfterUnExecute(){ _ieExecuted = false; }
    protected abstract void UnExecuteImpl();

    protected boolean BeforeDestroy() { return _isAlive; }
    protected void AfterDestroy(){ _isAlive = false; }
    protected abstract void DestroyImpl() throws Exception;

    private boolean _isAlive = true;
    private boolean _ieExecuted = false;
}
